modules/programs/fish.nix

```diff
@@ -138,9 +138,72 @@ let
     };
   };
 
-  abbrsStr = concatStringsSep "\n"
-    (mapAttrsToList (k: v: "abbr --add --global -- ${k} ${escapeShellArg v}")
-      cfg.shellAbbrs);
+  abbrModule = types.submodule {
+    options = {
+      expansion = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        description = ''
+          The command expanded by an abbreviation.
+        '';
+      };
+
+      position = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        example = "anywhere";
+        description = ''
+          If the position is "command", the abbreviation expands only if
+          the position is a command. If it is "anywhere", the abbreviation
+          expands anywhere.
+        '';
+      };
+
+      regex = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        description = ''
+          The regular expression pattern matched instead of the literal name.
+        '';
+      };
+
+      setCursor = mkOption {
+        type = with types; (either bool str);
+        default = false;
+        description = ''
+          The marker indicates the position of the cursor when the abbreviation
+          is expanded. When setCursor is true, the marker is set with a default
+          value of "%".
+        '';
+      };
+
+      function = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        description = ''
+          The fish function expanded instead of a literal string.
+        '';
+      };
+    };
+  };
+
+  abbrsStr = concatStringsSep "\n" (mapAttrsToList (name: def:
+    let
+      modifierStr = n: v: optional (v != null) ''--${n} "${toString v}"'';
+      modifierBool = n: v: optional (v != null && v) "--${n}";
+      modifierBoolOrStr = n: v:
+        if isBool v then
+          modifierBool n v
+        else
+          optional (v != null) "--${n}=${toString v}";
+      mods = with def;
+        modifierStr "position" position ++ modifierStr "regex" regex
+        ++ modifierBoolOrStr "set-cursor" setCursor
+        ++ modifierStr "function" function;
+      modifiers = if isAttrs def then toString mods else "";
+      expansion = if isAttrs def then def.expansion else def;
+    in "abbr --add ${modifiers} -- ${name}"
+    + optionalString (expansion != null) " \"${expansion}\"") cfg.shellAbbrs);
```

Comment:
This is reinventing too many wheels. At least `escapeShellArg`(`s`), maybe [`toGNUCommandLine`](https://github.com/NixOS/nixpkgs/blob/master/lib/cli.nix).

Reply:
Right so `toGNUCommandLine` is perhaps not the right tool here, but please at least generate a list of words and use `escapeShellArgs` on it.

Reply:
Thanks for the suggestion.
I rewrote the `modifiers` using `toGNUCommandLineShell`.
On the other hand, I left `expansion` as is because, in [some cases](https://github.com/nix-community/home-manager/pull/3806/files#diff-ab9e6378ee2cecd492130a366ed79158c145115d81928d86c93a2285fef1a2e1R27-R31), it did not work correctly when using `escapeShellArg`.

Reply:
You'd be reverting https://github.com/nix-community/home-manager/commit/a08dabf015081666f50ad1538f85d69e2217585e and maybe breaking @cole-h 's configuration.

In the case you're pointing to, it seems like

```nix
expansion = ''
  for dir in */
    cd $dir
    !
    cd ..
  end'';
```

would do. Can you actually think of use cases for allowing definition-time expansion in the `expansion`?

Reply:
I agree that definition-time expansion isn't necessary. `function` is a more flexible way to build abbrs from arbitrary code
```fish
functions.__4DIRS = "string join \\n -- 'for dir in */' 'cd $dir' '!' 'cd ..' 'end'";
shellAbbrs.4DIRS = {
  setCursor = "!";
  function = "__4DIRS";
};
```

If the function is slow and you don't mind slowing down shell startup instead you can still throw it in `interactiveShellInit` instead of using the abbr module. That's an inconvenience but imo it's outweighed by the convenience of not having to manually escape abbr bodies, and of course backwards compatibility.